test/apps/iodemo/io_demo.cc

```diff
@@ -681,52 +696,56 @@ class DemoClient : public P2pDemoCommon {
 
     DemoClient(const options_t& test_opts) :
         P2pDemoCommon(test_opts),
-        _num_sent(0), _num_completed(0), _status(OK), _start_time(get_time()),
-        _retry(0), _read_callback_pool(opts().iomsg_size)
-    {
-        _status_str[OK]                    = "ok";
-        _status_str[ERROR]                 = "error";
-        _status_str[RUNTIME_EXCEEDED]      = "run-time exceeded";
-        _status_str[CONN_RETRIES_EXCEEDED] = "connection retries exceeded";
+        _num_sent(0), _num_completed(0),
+        _status(OK), _start_time(get_time()),
+        _retry(0), _read_callback_pool(opts().iomsg_size) {
     }
 
     typedef enum {
         OK,
-        ERROR,
-        RUNTIME_EXCEEDED,
-        CONN_RETRIES_EXCEEDED
+        CONN_RETRIES_EXCEEDED,
+        WAIT_RESPONSES_TIMEOUT
     } status_t;
 
-    size_t do_io_read(UcxConnection *conn, uint32_t sn) {
-        size_t data_size = get_data_size();
-        IoMessage *m     = _io_msg_pool.get();
+    uint32_t get_server_index(const UcxConnection *conn) {
+        return _conn_idx[conn];
+    }
+
+    size_t do_io_read(int server_index, uint32_t sn) {
+        server_info_t& server_info = _server_info[server_index];
+        size_t data_size           = get_data_size();
+        IoMessage *m               = _io_msg_pool.get();
 
         m->init(IO_READ, sn, data_size, opts().validate);
-        if (!send_io_message(conn, m)) {
-            return data_size;
+        if (!send_io_message(server_info.conn, m)) {
+            return 0;
         }
 
+        ++server_info.num_sent;
         ++_num_sent;
 
         BufferIov *iov            = _data_buffers_pool.get();
         IoReadResponseCallback *r = _read_callback_pool.get();
 
         iov->init(data_size, _data_chunks_pool, sn, false);
-        r->init(&_num_completed, sn, opts().validate, iov);
+        r->init(&_num_completed, &server_info.num_completed[IO_READ], sn,
+                opts().validate, iov);
 
-        recv_data(conn, *iov, sn, r);
-        conn->recv_data(r->buffer(), opts().iomsg_size, sn, r);
+        recv_data(server_info.conn, *iov, sn, r);
```

Comment:
the same, `conn` can be failed above on `send_io_message` 
and the same for write

Reply:
conn can be deleted only from err callback which is called from progress